app/controllers/concerns/abstract_xikolo_connector.rb

```diff
@@ -49,4 +50,60 @@ def handle_enrollments_response(response_data, user)
     end
     evaluate_enrollments_update_map update_map, user
   end
+
+  def get_dates_for_user(user)
+    token_string = "Legacy-Token token=#{get_access_token user}"
+    api_url = self.class::ROOT_API_V2 + DATES_API
+    response = RestClient.get(api_url, accept: 'application/vnd.xikoloapplication/vnd.xikolo.v1, application/json', authorization: token_string)
+    JSON.parse response
+  end
+
+  def handle_dates_response(response_data, user)
+    update_map = create_update_map_for_user_dates user, mooc_provider
+    response_data['dates'].each do |response_user_date|
+      user_date = UserDate.find_by(user: user, mooc_provider: mooc_provider, ressource_id_from_provider: response_user_date['resource_id'], kind: response_user_date['kind'])
+      if user_date.present?
+        update_map[user_date.id] = true
+        update_existing_entry user_date, response_user_date
+      else
+        create_new_entry user, response_user_date
+      end
+    end
+    change_existing_no_longer_relevant_entries update_map
+  end
+
+  def create_new_entry(user, response_user_date)
+    user_date = UserDate.new
+    user_date.user = user
+    user_date.mooc_provider = mooc_provider
+    user_date.course = Course.find_by(provider_course_id: response_user_date['course_id'])
```

Comment:
What do you mean with this question: "Why are you using the mooc provider id above?"?
The proposed method I will use. If there is no course, it will be nil.